### ISSUE TITLE
Proposition: brackets for createQuery()

### ISSUE THREAD

**Issue by theUm** [graphaware/neo4j-php-ogm#129]
It seems that `createQuery()` is capable to support brackets in return statement (like in `RETURN { movies:  collect(MovieNode), actor: ActorNode}`). But node mappings do not becomes linked with nodes inside brackets, so they become simple `GraphAware\Common\Type\Node` instances.

Example - createQuery part:
```
MATCH ... (fv:FieldValue)-[:is_field_of]->(o:Object)
...
WITH o, {etf:etf.slug, nodes: collect(fv)} as fvs
	RETURN {o:o.name, nodes: collect(fvs)}
```
Mapping part:
```$query->addEntityMapping('fv', FieldValueNode::class);```

Output, notice `0 => Node {#1589` instead `FieldValueNode` node : 
```
"{o:o.name, nodes: collect(fvs)}" => array:2 [
        "o" => "THE_OBJECT"
        "nodes" => array:12 [
          0 => array:2 [
            "etf" => "verification"
            "nodes" => array:1 [
              0 => Node {#1589
                #identity: 1114
                #labels: array:1 []
                #properties: array:1 []
              }
            ]
          ]
	...
```

**Comment by ikwattro** [comment]
Yes, you actually return a map. Maps are not supported yet, however what you return is a map with nodes being a collection of maps. So might be hard to specify this in your mapping :/

**Comment by theUm** [comment]
Hi. Check out my `Query`'s maps support implementation.

Short explanation: maps in query returns as PHP assoc arrays, so my code just cycles through that array and seeks familiar array keys. to do so I created 2 additional constants HYDRATE_MAP, HYDRATE_MAP_COLLECTION.
Syntax for them are 
`addEntityMapping('%MAP_ALIAS%', null, Query::HYDRATE_MAP);`
for query like `{param1:node.val, param2: node}`
and
`addEntityMapping('%ARRAY_KEY%', null, Query::HYDRATE_MAP_COLLETION);`
for query like 
`WITH node, { otherNode.name, otherNode.fieldName} as cols RETURN {val1:value, data: collect(cols) }`

I`m not heavily tested it, ( works for my my case well (but not fast) (as expected for recursive array crawler) ) so I`ll catch up to this later

**Comment by ikwattro** [comment]
Thanks @theUm 

Can you add a minimal example ? Like entity , the query used and the results you get. 

No need for adding it in the PR, just paste as comment;

Thx

**Comment by theUm** [comment]
Entities:
```
<?php

namespace BuilderBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use GraphAware\Neo4j\OGM\Annotations as OGM;

/**
 * @OGM\Node(label="FieldValue")
 */
class FieldValueNode
{
//    use FileFieldTrait;

    /**
     * @OGM\GraphId()
     * @var int
     */
    protected $id;

    /**
     * @OGM\Property(type="string", key="data", nullable=true)
     * @var string
     */
    protected $data = null;

    /**
     * @OGM\Relationship(type="is_value_of", direction="OUTGOING", targetEntity="TypeFieldNode", collection=false)
     * @var TypeFieldNode
     */
    protected $typeField;

    /**
     * @OGM\Relationship(type="is_field_of", direction="OUTGOING", targetEntity="ObjectNode", collection=true)
     * @var ArrayCollection|ObjectNode[]
     */
    protected $objects;

    /**
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }
}
```
```
<?php

namespace BuilderBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use GraphAware\Neo4j\OGM\Annotations as OGM;

/**
 * @OGM\Node(label="FieldValue")
 */
class FieldValueNode
{
//    use FileFieldTrait;

    /**
     * @OGM\GraphId()
     * @var int
     */
    protected $id;

    /**
     * @OGM\Property(type="string", key="data", nullable=true)
     * @var string
     */
    protected $data = null;

    /**
     * @OGM\Relationship(type="is_value_of", direction="OUTGOING", targetEntity="TypeFieldNode", collection=false)
     * @var TypeFieldNode
     */
    protected $typeField;

    /**
     * @OGM\Relationship(type="is_field_of", direction="OUTGOING", targetEntity="ObjectNode", collection=true)
     * @var ArrayCollection|ObjectNode[]
     */
    protected $objects;

    /**
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }
}
```

Query code:
```
        $entityTypeFieldsQuery = $nm->createQuery(
            'MATCH (etf:EntityTypeField)-[:has_field]->(type:EntityType)<-[:has_type]-(o:Object)
            WHERE o.slug = {oSlug}
            OPTIONAL MATCH (etf)<-[:is_value_of]-(fv)-[:is_field_of]->(o)
            WITH o, {etf:etf.slug, fields: collect(fv)} as fieldValues
            RETURN {o:o, fieldsByTypes: collect(fieldValues)} as objectStruct'
        );
        $entityTypeFieldsQuery->addEntityMapping('objectStruct', null, Query::HYDRATE_MAP);
        $entityTypeFieldsQuery->addEntityMapping('o', ObjectNode::class);
        $entityTypeFieldsQuery->addEntityMapping('fieldsByTypes', null, Query::HYDRATE_MAP_COLLETION);
        $entityTypeFieldsQuery->addEntityMapping('fields', FieldValueNode::class, Query::HYDRATE_COLLECTION);
        $entityTypeFieldsQuery->setParameter('oSlug', 'tickmill-armada-markets');
        $res = $entityTypeFieldsQuery->execute();
        VarDumper::dump($res);
        die;
```

Result:
![res](https://cloud.githubusercontent.com/assets/3176889/25010705/966bde0c-2073-11e7-97ed-ca0c1c36b969.png)

Sample data for neo4j:
```
begin
commit
begin
create (_1131:`EntityTypeField` {`fieldType`:"integer", `hasOwnUrl`:false, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"Год выхода на рынок", `slug`:"established-date", `tabGroup`:"Рег данные", `tooltip`:""})
create (_1142:`EntityType` {`description`:"", `isDataType`:false, `name`:"Брокеры", `slug`:"brokers"})
create (_1200:`Object` {`description`:"", `name`:"Tickmill (Armada Markets)", `slug`:"tickmill-armada-markets"})
create (_1239:`EntityTypeField` {`fieldType`:"predefSelect2", `hasOwnUrl`:false, `isCollection`:true, `isMainField`:false, `metaDesc`:"", `name`:"Категории брокеров", `predefinedFields`:["Форекс-брокер", "CFD-брокер", "Фьючерсный бокер", "Акционный брокер", "Опционный брокер", "Спред-беттинг брокер", "Парного трейдинга брокер", "Бинарных опционов брокер", "Криптовалютный брокер"], `slug`:"broker-category", `tabGroup`:"Разделы брокера", `tooltip`:""})
create (_1287:`FieldValue` {`data`:"США"})
create (_1326:`EntityTypeField` {`fieldType`:"text", `hasOwnUrl`:true, `isCollection`:true, `isMainField`:false, `metaDesc`:"", `name`:"Ограничения по странам", `slug`:"country-restrictions", `tabGroup`:"Ограничения", `tooltip`:""})
create (_1328:`EntityTypeField` {`fieldType`:"checkbox", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"Вывести рекомендованных", `slug`:"show-recom", `tabGroup`:"Ограничения", `tooltip`:"Заглушить переход на сайт брокера"})
create (_1330:`EntityTypeField` {`fieldType`:"date", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"Дата закрытия брокера", `slug`:"broker-close-date", `tabGroup`:"Ограничения", `tooltip`:"Если брокер закрыт"})
create (_1335:`EntityTypeField` {`fieldType`:"text", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"Адрес головного офиса", `slug`:"head-office", `tabGroup`:"Рег данные", `tooltip`:""})
create (_1336:`FieldValue` {`data`:"Trop-X Securities Exchange building, 3 F28-F29 Eden Plaza, Eden Island, Mahe, Republic of Seychelles"})
create (_1337:`EntityTypeField` {`fieldType`:"text", `hasOwnUrl`:true, `isCollection`:true, `isMainField`:false, `metaDesc`:"", `name`:"Отрицательные стороны", `slug`:"negaties", `tabGroup`:"Разделы брокера", `tooltip`:""})
create (_1338:`EntityTypeField` {`fieldType`:"text", `hasOwnUrl`:true, `isCollection`:true, `isMainField`:false, `metaDesc`:"", `name`:"Положительные стороны", `slug`:"positives", `tabGroup`:"Разделы брокера", `tooltip`:""})
create (_1382:`FieldValue` {`data`:"Форекс-брокер"})
create (_1383:`FieldValue` {`data`:"CFD-брокер"})
create (_1549:`FieldValue` {`data`:2015})
create (_1667:`EntityTypeField` {`fieldType`:"simple_text", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"SEO title", `slug`:"seo-title", `tabGroup`:"SEO", `tooltip`:""})
create (_1668:`EntityTypeField` {`fieldType`:"textArea", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"SEO description", `slug`:"seo-descr", `tabGroup`:"SEO", `tooltip`:""})
create (_1669:`EntityTypeField` {`fieldType`:"text", `hasOwnUrl`:true, `isCollection`:true, `isMainField`:false, `metaDesc`:"", `name`:"SEO meta tags", `slug`:"seo-tags", `tabGroup`:"SEO", `tooltip`:""})
create (_1671:`FieldValue` {`data`:"Armada Markets"})
create (_1672:`FieldValue` {`data`:"отзывы брокеры форекс"})
create (_1673:`FieldValue` {`data`:"брокер форекс"})
create (_1693:`FieldValue` {`data`:"Обзор брокера форекс TickMill (Armada Markets). Отзывы, опыт торговли в ТикМилл"})
create (_1759:`EntityTypeField` {`fieldType`:"checkbox", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"Черный список брокеров", `slug`:"brokers-blacklist", `tabGroup`:"Ограничения", `tooltip`:""})
create (_1760:`EntityTypeField` {`fieldType`:"checkbox", `hasOwnUrl`:true, `isCollection`:false, `isMainField`:false, `metaDesc`:"", `name`:"Сомнительные брокеры", `slug`:"watch-list", `tabGroup`:"Ограничения", `tooltip`:""})
create (_1131)-[:`has_field`]->(_1142)
create (_1200)-[:`has_type`]->(_1142)
create (_1239)-[:`has_field`]->(_1142)
create (_1287)-[:`is_value_of`]->(_1326)
create (_1287)-[:`is_field_of`]->(_1200)
create (_1326)-[:`has_field`]->(_1142)
create (_1328)-[:`has_field`]->(_1142)
create (_1330)-[:`has_field`]->(_1142)
create (_1335)-[:`has_field`]->(_1142)
create (_1336)-[:`is_field_of`]->(_1200)
create (_1336)-[:`is_value_of`]->(_1335)
create (_1337)-[:`has_field`]->(_1142)
create (_1338)-[:`has_field`]->(_1142)
create (_1382)-[:`is_value_of`]->(_1239)
create (_1382)-[:`is_field_of`]->(_1200)
create (_1383)-[:`is_field_of`]->(_1200)
create (_1383)-[:`is_value_of`]->(_1239)
create (_1549)-[:`is_value_of`]->(_1131)
create (_1549)-[:`is_field_of`]->(_1200)
create (_1667)-[:`has_field`]->(_1142)
create (_1668)-[:`has_field`]->(_1142)
create (_1669)-[:`has_field`]->(_1142)
create (_1671)-[:`is_field_of`]->(_1200)
create (_1671)-[:`is_value_of`]->(_1669)
create (_1672)-[:`is_value_of`]->(_1669)
create (_1672)-[:`is_field_of`]->(_1200)
create (_1673)-[:`is_field_of`]->(_1200)
create (_1673)-[:`is_value_of`]->(_1669)
create (_1693)-[:`is_field_of`]->(_1200)
create (_1693)-[:`is_value_of`]->(_1667)
create (_1759)-[:`has_field`]->(_1142)
create (_1760)-[:`has_field`]->(_1142)
;
commit

```

**Comment by theUm** [comment]
Further tests are required. Tommorow or maybe at holidays I`ll test diffrent queries with this

**Comment by ikwattro** [comment]
Thanks a lot @theUm .

I will take it a spin as well. Happy to get more use cases to cover. 

Maybe other users could give as well their opinion ?

**Comment by ikwattro** [comment]
@theUm 

A quick question : 

```
MATCH (etf:EntityTypeField)-[:has_field]->(type:EntityType)<-[:has_type]-(o:Object)
            WHERE o.slug = {oSlug}
            OPTIONAL MATCH (etf)<-[:is_value_of]-(fv)-[:is_field_of]->(o)
            WITH o, {etf:etf.slug, fields: collect(fv)} as fieldValues
            RETURN {o:o, fieldsByTypes: collect(fieldValues)} as objectStruct
```

Why do you return a map, and not simply : 

```
MATCH (etf:EntityTypeField)-[:has_field]->(type:EntityType)<-[:has_type]-(o:Object)
            WHERE o.slug = {oSlug}
            OPTIONAL MATCH (etf)<-[:is_value_of]-(fv)-[:is_field_of]->(o)
            RETURN o, {etf:etf.slug, fields: collect(fv)} as fieldValues
```

?

**Comment by theUm** [comment]
Uh, yes its better solution, my bad. Thanks for improvement. However, in some more complex cases, I will need query like provided (like as fetching all `:FieldValues` of `:FieldTypes` of all `:Objects` of specific `:EntityType`). It will be not widely used query, but better be ready for that situation as I predict I will be encountering it soon 😃
BTW, thanks for your work again, I understand what is like to trying to get people to use something new and unfamiliar 😃

**Comment by ikwattro** [comment]
Fixed by #131

**Comment by theUm** [comment]
Thank you
